src/app/util/af-types.h

```diff
@@ -131,6 +131,16 @@ typedef struct
     const chip::CommandId * generatedCommandList;
 } EmberAfCluster;
 
+/**
+ * @brief Struct that represents a logical device type consisting
```

Comment:
Why not just use `Descriptor::Structs::DeviceType::Type` directly, instead of defining a different struct that needs to be converted to it?

Reply:
I do not think core data model libraries should be depending on code-generated schema bits - that dependency should only be happening the level above this, i.e the cluster logic and application bits.

The schema should be thought of as one projection of this core state and as such, the API here shouldn't be coupled to how we choose to present/project this information onto a cluster representation.

Reply:
> I do not think core data model libraries should be depending on code-generated schema bits

The core data model libraries depend on all sorts of code-generated stuff: enum values, etc, etc.

The only outcome of not depending on the exact same struct here, is that we have to write glue code to convert one to the other.

Yes, if the cluster representation were not suitable we would do that.  But in this case it's fine; why not use it?

Reply:
> The core data model libraries depend on all sorts of code-generated stuff: enum values, etc, etc.

There is a distinction between depending on code-generated data from schema but not actually depending on a specific type defined in a cluster (e.g the `GENERATED_ATTRIBUTES`), and depending on a specific type definition as defined in a cluster. This latter bit I don't believe we do in the data model libraries. The danger with doing so is that the API contract defined by the data model library is now a function of a schema definition. If we decide to change the schema definition in the future to alter that cluster, the API contract changes as well, which is not desirable.

The other thing of note is that depending on generated content in CPP files is fine. It really should be frowned upon in header files in data model library code. In  the existing header files in `app/util`, there is only one file that actually depends on generated content (`types_stub.h`). I intend to fix that to not do that, and shift that up to actual cluster logic (have a PR ready for that as well soon).

I'd very much like to avoid injecting a new header-file depending on `cluster-objects.h` if we could. 

Perhaps the TL;DR here for me is less about the specific instance here of device types (which perhaps, might not change a whole lot and looks terribly convenient to use), and the danger and precedence I really want to avoid setting.

Reply:
> If we decide to change the schema definition in the future to alter that cluster

Cluster enum values would not be changed in a backward-incompatible way, though.

> I'd very much like to avoid injecting a new header-file depending on cluster-objects.h if we could.

This is why `cluster-enums.h` is a separate header.

Reply:
Chatted with @bzbarsky-apple offline, and we agreed to keep the pattern as is in this PR since there are non-zero public APIs involved here. If this becomes onerous in the future, we can re-visit this.